plugins/inputs/vsphere/endpoint.go

```diff
@@ -0,0 +1,745 @@
+	ts     time.Time
+	fields map[string]interface{}
+}
+
+type objectMap map[string]objectRef
+
+type objectRef struct {
+	name      string
+	ref       types.ManagedObjectReference
+	parentRef *types.ManagedObjectReference //Pointer because it must be nillable
+	guest     string
+}
+
+type resourceInfo struct {
+	name    string
+	metrics performance.MetricList
+}
+
+type metricQResponse struct {
+	obj     objectRef
+	metrics *performance.MetricList
+}
+
+// NewEndpoint returns a new connection to a vCenter based on the URL and configuration passed
+// as parameters.
+func NewEndpoint(parent *VSphere, url *url.URL) *Endpoint {
+	e := Endpoint{
+		pool:         Pool{u: url, v: parent, root: nil},
+		URL:          url,
+		Parent:       parent,
+		lastColls:    make(map[string]time.Time),
+		instanceInfo: make(map[string]resourceInfo),
+		initialized:  false,
+	}
+
+	e.resourceKinds = map[string]resourceKind{
+		"cluster": {
+			pKey:             "clustername",
+			enabled:          parent.GatherClusters,
+			realTime:         false,
+			sampling:         300,
+			objects:          make(objectMap),
+			includes:         parent.ClusterMetricInclude,
+			excludes:         parent.ClusterMetricExclude,
+			collectInstances: parent.ClusterInstances,
+			getObjects:       getClusters,
+		},
+		"host": {
+			pKey:             "esxhostname",
+			enabled:          parent.GatherHosts,
+			realTime:         true,
+			sampling:         20,
+			objects:          make(objectMap),
+			includes:         parent.HostMetricInclude,
+			excludes:         parent.HostMetricExclude,
+			collectInstances: parent.HostInstances,
+			getObjects:       getHosts,
+		},
+		"vm": {
+			pKey:             "vmname",
+			enabled:          parent.GatherVms,
+			realTime:         true,
+			sampling:         20,
+			objects:          make(objectMap),
+			includes:         parent.VmMetricInclude,
+			excludes:         parent.VmMetricExclude,
+			collectInstances: parent.VmInstances,
+			getObjects:       getVMs,
+		},
+		"datastore": {
+			pKey:             "dsname",
+			enabled:          parent.GatherDatastores,
+			realTime:         false,
+			sampling:         300,
+			objects:          make(objectMap),
+			includes:         parent.DatastoreMetricInclude,
+			excludes:         parent.DatastoreMetricExclude,
+			collectInstances: parent.DatastoreInstances,
+			getObjects:       getDatastores,
+		},
+	}
+
+	return &e
+}
+
+func (e *Endpoint) init() error {
+
+	err := e.setupMetricIds()
+	if err != nil {
+		log.Printf("E! Error in metric setup for %s: %v", e.URL.Host, err)
+		return err
+	}
+
+	if e.Parent.ObjectDiscoveryInterval.Duration.Seconds() > 0 {
+		discoverFunc := func() {
+			err = e.discover()
+			if err != nil {
+				log.Printf("E! Error in initial discovery for %s: %v", e.URL.Host, err)
+			}
+
+			// Create discovery ticker
+			//
+			e.discoveryTicker = time.NewTicker(e.Parent.ObjectDiscoveryInterval.Duration)
+			go func() {
+				for range e.discoveryTicker.C {
```

Comment:
Good point. I'm struggling to find a way to close it on exit. There's no "teardown" or similar on the Input interface. How can I sense when the plugin is about to be reloaded/destroyed so I can kill the goroutine?

Reply:
You will probably need to use the ServiceInput interface so you can use the Start()/Stop() function.

Reply:
Thanks. How much time can I spend in a Stop()? Are you expecting it to return more or less immediately or can I wait for things to shut down gracefully? As you've probably noticed, a lot of processing happens concurrently and asynchronously using the WorkerPool construct. I've implemented a simple graceful shutdown function (not yet pushed as of writing this) that's essentially just an atomic flag that's checked every time the worker pulls a workitem from the channel. If the flag is set, it simply bails out gracefully.

The question is: Should I wait for all work to finish before I return from Stop() or should I just rely on it finishing in the background? It may take a few seconds, maybe up to 30s in extreme cases. What's the best practice?

Reply:
It is best to wait in Stop, you might want to consider using a `context.Context` to stop the goroutines, the advantage of this is that you can select on it.

Reply:
*slaps forehead* Context would be a much better design, yes! Redesigned to have a root context that can cancel any outstanding requests. Stand by for push!

Reply:
Done. Using Context throughout the code. Especially neat, since govmomi supports Context and allows us to immediately abort long running calls.